### PR TITLE
Add Policies

### PR DESCRIPTION
**Pull request by chx2**

### CODE REVIEW

**Comment by chx2** [comment]
@drewroberts before I go back through and refactor all the ones I've worked on, does this package now reflect the format for policies correctly?

**Comment by drewroberts** [comment]
I think it has everything. I'll let @pdbreen & @joshtorres take a look at it too.

File: database/seeders/PermissionsSeeder.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+use Carbon\Carbon;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Database\Seeder;
+use Illuminate\Support\Facades\Schema;
+
+class PermissionsSeeder extends Seeder
+{
+    public function run()
+    {
+        if (Schema::hasTable('permissions')) {
```
Comment (sl0wik) [inline database/seeders/PermissionsSeeder.php]: Is this necessary?

**Comment by pdbreen** [comment]
Code looks fine -- need to add the env values to `phpunit.xml.dist` in order for the tests to work properly.  There is also a sample tests\Unit\Seeders\PermissionsSeeder.php test in discounts/vouchers/checkout that could be added in.
```
    <php>
        <env name="DB_CONNECTION" value="testing"/>
        <env name="APP_KEY" value="AckfSECXIvnK5r28GVIWUAxmbBSjTsmF"/>
    </php>
```

File: database/seeders/PermissionsSeeder.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            $now = Carbon::now()->format('Y-m-d H:i:s');
+            $permissions = [
+                'view signatures',
+                'create signatures',
+                'update signatures',
+                'delete signatures'
+            ];
+
+            DB::table('permissions')->insertOrIgnore(
```
Comment (sl0wik) [inline database/seeders/PermissionsSeeder.php]: I would avoid calling database directly, cant we use `app('permission')` or something?

File: database/seeders/PermissionsSeeder.php (hunk 1 above)
Comment (pdbreen) [reply database/seeders/PermissionsSeeder.php]: If use of permissions is to be optional, then yes this is required -- in theory, this project could be used by some other project that doesn't have permissions.  If we make use of permissions an explicit dependency, then these seeders will be redone using `Permissions` contract methods.

File: database/seeders/PermissionsSeeder.php (hunk 2 above)
Comment (pdbreen) [reply database/seeders/PermissionsSeeder.php]: Permissions is not our model.

Comment (drewroberts) [reply database/seeders/PermissionsSeeder.php]: We could extend it from that Spatie package and have it be our own model in the tger/authorization package

Comment (pdbreen) [reply database/seeders/PermissionsSeeder.php]: Not necessary.   Once the spatie package is a dependency, the seeder implementation is entirely different.  I'll illustrate in a PR on vouchers.

Comment (pdbreen) [reply database/seeders/PermissionsSeeder.php]: https://github.com/tipoff/vouchers/pull/18

File: tests/Unit/Policies/SignaturePolicyTest.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+     * @test
+     * @dataProvider data_provider_for_all_permissions_as_creator
+     */
+    public function all_permissions_as_creator(string $permission, UserInterface $user, bool $expected)
+    {
+        $discount = Signature::factory()->make([
+            'creator_id' => $user,
+        ]);
+
+        $this->assertEquals($expected, $user->can($permission, $discount));
```
Comment (drewroberts) [inline tests/Unit/Policies/SignaturePolicyTest.php]: This needs to be changed from discount to signature.

File: tests/Unit/Policies/SignaturePolicyTest.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * @test
+     * @dataProvider data_provider_for_all_permissions_not_creator
+     */
+    public function all_permissions_not_creator(string $permission, UserInterface $user, bool $expected)
+    {
+        $discount = Signature::factory()->make();
+
+        $this->assertEquals($expected, $user->can($permission, $discount));
```
Comment (drewroberts) [inline tests/Unit/Policies/SignaturePolicyTest.php]: Same here.

**Comment by chx2** [comment]
Removed the policy test for now. Will need to figure out tests for packages that don't directly have a user associated them via `creator_id`

**Review by drewroberts — APPROVED** [review]
Thank you, Chris! 🏋️

**Pull request merged**